grapher/text/MarkdownTextWrap.tsx

```diff
@@ -0,0 +1,510 @@
+    return tokens.map((t) => t.toPlaintext()).join("")
+}
+
+export function splitIntoLines(
+    tokens: IRToken[],
+    maxWidth: number
+): IRToken[][] {
+    const processedLines: IRToken[][] = []
+    const unprocessedLines: IRToken[][] = splitAllOnNewline(tokens)
+    while (unprocessedLines.length) {
+        const currentLine = unprocessedLines.shift()!
+        if (getLineWidth(currentLine) <= maxWidth) {
+            processedLines.push(currentLine)
+        } else {
+            const breakpoint = getBreakpointBefore(currentLine, maxWidth)
+            if (!breakpoint) {
+                processedLines.push(currentLine)
+            } else {
+                const { before, after } = splitLineAtBreakpoint(
+                    currentLine,
+                    breakpoint.breakOffset
+                )
+                processedLines.push(before)
+                unprocessedLines.unshift(after)
+            }
+        }
+    }
+    return processedLines
+}
+
+export function parsimmonToTextTokens(
+    nodes: EveryMarkdownNode[],
+    fontParams?: IRFontParams
+): IRToken[] {
+    return nodes.map((node): IRToken => {
+        if (node.type === "text") {
+            return new IRText(node.value, fontParams)
+        } else if (node.type === "newline") {
+            return new IRLineBreak()
+        } else if (node.type === "whitespace") {
+            return new IRWhitespace(fontParams)
+        } else if (
+            node.type === "bold" ||
+            node.type === "plainBold" ||
+            node.type === "boldWithoutItalic"
+        ) {
+            return new IRBold(
+                parsimmonToTextTokens(node.children, {
+                    ...fontParams,
+                    fontWeight: 700,
+                })
+            )
+        } else if (
+            node.type === "italic" ||
+            node.type === "plainItalic" ||
+            node.type === "italicWithoutBold"
+        ) {
+            return new IRItalic(
+                parsimmonToTextTokens(node.children, {
+                    ...fontParams,
+                    isItalic: true,
+                })
+            )
+        } else if (node.type === "plainUrl") {
+            return new IRLink(
+                node.href,
+                parsimmonToTextTokens(
+                    [{ type: "text", value: node.href }],
+                    fontParams
+                )
+            )
+        } else if (node.type === "markdownLink") {
+            return new IRLink(
+                node.href,
+                parsimmonToTextTokens(node.children, fontParams)
+            )
+        } else if (node.type === "detailOnDemand") {
+            // TODO: create a IR for this
+            return new IRLink(
+                node.term,
+                parsimmonToTextTokens(node.children, fontParams)
+            )
+        } else {
+            throw new Error(`Unknown node type: ${(node as any).type}`)
+        }
```

Comment:
Fine like this but just as an FYI - the ts-pattern library that does pattern matching with exhaustiveness checking can be pretty neat for such cases (would need a union type that merges all possible node types but that might be a good idea anyhow). Nothing to here, just wanted to mention this.